include/ozo/libpq.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include <ozo/type_traits.h>
+
+#include <libpq-fe.h>
+
+#include <boost/iterator/iterator_facade.hpp>
+#include <memory>
+#include <vector>
+
+
+namespace ozo::libpq {
+
+class value {
+public:
+    value(PGresult* r, int row, int col) : r_(r), row_(row), col_(col) {}
+
+    inline oid_t oid() const { return PQftype(r_, col_); }
+    inline const char* bytes() const { return PQgetvalue(r_, row_, col_); }
+    inline int size() const { return PQgetlength(r_, row_, col_); }
+
+private:
+    PGresult* r_;
+    int row_;
+    int col_;
+};
+
+
+class row {
+public:
+    class iterator : public boost::iterator_facade<
+        iterator,
+        value,
+        boost::forward_traversal_tag, // maybe will implement random access later
+        const value&,
+        int
+    > {
+    public:
+        iterator() = default;
+        iterator(PGresult* r, int row, int col) : r_(r), row_(row), col_(col),
+            v_(r, row, col) {}
+
+    private:
+        const value& dereference() const {
+            return v_;
+        }
+
+        bool equal(const iterator& rhs) const {
+            return r_ == rhs.r_ && row_ == rhs.row_ && col_ == rhs.col_;
+        }
+
+        void increment() {
+            if (col_ + 1 < PQnfields(r_)) {
+                ++ col_;
+                v_ = {r_, row_, col_};
+            } else {
+                *this = {};
+            }
+        }
+
+        PGresult* r_ {nullptr};
+        int row_ {0};
+        int col_ {0};
+
+        value v_ {nullptr, 0, 0};
+
+        friend class boost::iterator_core_access;
+    };
+
+    row() = default;
+    row(PGresult* r, int row) : r_(r), row_(row) {}
+
+    iterator begin() const { return iterator{r_, row_, 0}; }
+    iterator end() const { return iterator{}; }
+
+    // TODO: replace usage of these in impl::row with a random access row iterator
+    std::size_t size() { return PQnfields(r_); }
+    value at(int i) { return value{r_, row_, i}; }
+
+private:
+    PGresult* r_ {nullptr};
+    int row_ {0};
+};
+
+
+class result
+{
+public:
+    class iterator : public boost::iterator_facade<
+        iterator,
+        row,
+        boost::forward_traversal_tag, // can be random_access, but we do not need it
```

Comment:
Why not?

Reply:
it's unnecessary for conversion as implemented now, because generic detail::result iterates forwards through libpq::result rows. If we decide to make detail::result (and thus ozo::result) a full-fledged container with random access, then we'll have to implement random access here.